adam/learner/pursuit.py

```diff
@@ -657,16 +637,230 @@ def compute_hypothesis_id(h: PerceptionGraphPattern) -> str:
                 ),
             )
 
+    @abstractmethod
+    def _find_identical_hypothesis(
+        self, new_hypothesis: HypothesisT, candidates: Iterable[HypothesisT]
+    ) -> Optional[HypothesisT]:
+        """
+        Finds the first hypothesis object, if any, in *candidates*
+        which is isomorphic to *new_hypothesis*.
+        """
+
+    @abstractmethod
+    def _candidate_perceptions(self, observed_perception_graph) -> List[PerceptionGraph]:
+        """
+        Returns a sequence of candidate perceptions of a specific type, given an *observed_perception_graph*.
+        """
+
+    @abstractmethod
+    def _candidate_meanings(
+        self, observed_perception_graph: PerceptionGraph
+    ) -> Sequence[HypothesisT]:
+        """
+        Returns a sequence of candidate meanings of a specific type, given a sequence of *perceptions*.
+        """
+
+    @abstractmethod
+    def _meaning_from_perception(self, perception: PerceptionGraph) -> HypothesisT:
+        """
+        Returns a meaning representation for a given *perception*.
+        """
+
+    @abstractmethod
+    def _compute_match_ratio(
+        self, hypothesis: HypothesisT, graph: PerceptionGraph
+    ) -> "PursuitLanguageLearner.PartialMatch":
+        """
+        Computes the fraction of pattern graph nodes of *hypothesis* which match *graph*.
+        """
+
+    @abstractmethod
+    def _check_hypothesis_isomorphism(
+        self, h: HypothesisT, hypothesis: HypothesisT
+    ) -> bool:
+        """
+        Checks if two hypotheses are isomorphic.
+        """
+
+    @abstractmethod
+    def _matcher(
+        self, hypothesis: HypothesisT, observed_perception_graph: PerceptionGraph
+    ) -> PatternMatching:
```

Comment:
Do we need to return `PatternMatching` here or can we return something more abstract? What properties of the returned object do we exploit?

Reply:
We utilize PatternMatching in the describe() method, I had to create an abstract method to modularize it such that it works with different HypothesisT types. I'm not sure about the alternatives, but we could potentially return the result of PatternMatching.matches() and change describe() slightly?